### ISSUE TITLE
Need help with generics

### ISSUE THREAD

**Issue by valkyrienyanko** [Valks-Games/sankari#248]
## Why do this?
To remove `public Slime Slime { get; set; }`

## The Steps to get to the Problem

If you changed `SlimeAnimationIdle.cs` to look like this.

![Capture](https://user-images.githubusercontent.com/6277739/207747657-6392e9ff-e67f-4952-bdae-d4c119b5e6bd.PNG)

This would complain in `Slime.cs`

It's hard to see in the screenshot but it says `CS0029 Cannot implicitly convert type 'Sankari.SlimeAnimationIdle' to 'Sankari.EntityAnimation<Sankari.MovingEntity>'`

![Untitled](https://user-images.githubusercontent.com/6277739/207747855-6fe6709b-04c5-48e2-a55d-f5360f1903af.png)

Wouldn't this mean `MovingEntity` would have to become `MovingEntity<T>` 

![Capture](https://user-images.githubusercontent.com/6277739/207748330-a3f6ff01-004b-4abe-a5a0-560c71f3d44d.PNG)

So this
```cs
public Dictionary<EntityAnimationType, EntityAnimation<MovingEntity>> Animations { get; set; } = new();
```
becomes this
```cs
public Dictionary<EntityAnimationType, EntityAnimation<T>> Animations { get; set; } = new();
```

Also this would have to change.

Before:
![Capture](https://user-images.githubusercontent.com/6277739/207748537-6dab60a6-a229-4b82-a344-57a1a620c38b.PNG)

After:
![Capture](https://user-images.githubusercontent.com/6277739/207748612-1f72df94-cb57-4a2c-ba2c-9c599e5ba2df.PNG)

## The Problem
And since `MovingEntity<T>` is now a generic thus `Slime` would have to become `Slime<T>` but what would you pass in for `<T>` for a slime?

This is where I get super confused.

![Capture](https://user-images.githubusercontent.com/6277739/207748702-445e8f17-123f-4e8b-ab4d-d4cf113837f3.PNG)

![Capture](https://user-images.githubusercontent.com/6277739/207748968-29145b0a-1855-4c4e-90d3-7ac037ed4073.PNG)

![Capture](https://user-images.githubusercontent.com/6277739/207749044-177d6c30-5562-4d0e-b7c0-07ed4cbb938b.PNG)

![Capture](https://user-images.githubusercontent.com/6277739/207749103-a94de7eb-b5b9-4ab2-82b9-13eec581bc60.PNG)

![Capture](https://user-images.githubusercontent.com/6277739/207749258-951627a8-35e1-4a14-a861-023bef970024.PNG)

**Comment by cathei** [comment]
What you want is [covariance](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/).

In C# covariance generic argument is supported for interfaces, not classes.

So this is not allowed:
```cs
EntityAnimation<MovingEntity> variable = new EntityAnimation<Slime>();
```

But if you make an interface support covariance such as:
```cs
interface IEntityAnimation<out T> { }

class EntityAnimation<T> : IEntityAnimation<T> { }
```

Then this is allowed:
```cs
IEntityAnimation<MovingEntity> variable = new EntityAnimation<Slime>();
```

**Comment by valkyrienyanko** [comment]
Thank you for this. I've decided I'm going to remove all generics where covariance is involved. I just tried to implement what you said but it felt like the codebase was getting even messier and I ended up not getting it to work due to my lack of experience. Looks like generics can only do so much. I'm going to keep this issue up until I've finished removed all the generics.

**Comment by valkyrienyanko** [comment]
Okay I resolved this issue thanks to some very helpful individuals. Cheers!

![Capture](https://user-images.githubusercontent.com/6277739/207788431-515ecb11-7185-4b14-b108-4a42af9855de.PNG)

Commit that resolves this: 190aac79cfc4587fffcf21499538da95fc4e60d4
